client/pkg/session/session.go

```diff
@@ -0,0 +1,176 @@
+package session
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"os"
+	"sync/atomic"
+
+	"github.com/wandb/wandb/client/internal/connection"
+	"github.com/wandb/wandb/client/internal/launcher"
+	"github.com/wandb/wandb/core/pkg/observability"
+	"github.com/wandb/wandb/core/pkg/service"
+)
+
+const (
+	localHost = "127.0.0.1"
+)
+
+// SessionParams holds parameters for creating a new Session
+type SessionParams struct {
+	CorePath string
+}
+
+// Session manages the lifecycle of a connection session
```

Comment:
it's not immediately clear to me that this refers to a connection to the wandb-core internal process

> Session is a connection to the internal "core" process.

also, maybe it would be more informative to call this `InternalProcess` instead of `Session`. Each "session" corresponds to one process, but the naming leaves open the possibility that there could be multiple sessions per process.

Reply:
Elaborated what connection we're referring to in the docstring.

As for the naming, I think Session is more appropriate. We are making it public with the intent to allow Go users to connect to a wandb-core service already running either locally or on the network.

Reply:
> the intent to allow Go users to connect to a wandb-core service already running

I see, that makes sense. It has a `Start()` method that launches a new internal process; I suppose we'd eventually add a `Connect()` method?

What if a user wants to start a process without connecting to it, with the intention to connect from other processes? It seems there's an opportunity to separate concepts:

* `WandbProcess` is a reference to a running process.
  * `WandbProcess::Connect() Session` opens and returns a connection to the process.
  * `WandbProcess::Kill()` kills the process.
* `Session` is a connection to a running W&B process.
    * `Session::Close()` closes the connection and frees any resources.
    * Other methods for communicating with the session.

These should be interfaces to make code that uses them testable.

Optionally, we can have `WandbProcessLauncher` and `SessionStarter`, but I think we can get away with top-level `Launch(settings) WandbProcess` and `Connect(address) Session` functions.

The simplest usage in Go would be:

```go
process := wandb.Launch(...)
defer process.Kill()

session := process.Connect()
defer session.Close()

session.Send(...)
```

Or to connect to a process running elsewhere:

```go
session := wandb.Connect(...)
defer session.Close()

session.Send(...)
```